File: frontend/src/api/swissRulesWatch.ts
import { apiClient } from '@/api/client';

export type FreshnessState = 'current' | 'aging' | 'stale' | 'unknown';
export type WatchTier = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export interface RuleSource {
  id: string;
  source_code: string;
  source_name: string;
  source_url: string | null;
  watch_tier: WatchTier;
  last_checked_at: string | null;
  last_changed_at: string | null;
  freshness_state: FreshnessState;
  change_types_detected: string[] | null;
  is_active: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface RuleChangeEvent {
  id: string;
  source_id: string;
  event_type: string;
  title: string;
  description: string | null;
  impact_summary: string | null;
  detected_at: string;
  reviewed: boolean;
  reviewed_by_user_id: string | null;
  reviewed_at: string | null;
  review_notes: string | null;
  affects_buildings: boolean;
  created_at: string;
}

export interface CommunalAdapter {
  id: string;
  commune_code: string;
  commune_name: string;
  canton_code: string;
  adapter_status: string;
  supports_procedure_projection: boolean;
  supports_rule_projection: boolean;
  fallback_mode: string;
  source_ids: string[] | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CommunalOverride {
  id: string;
  commune_code: string;
  canton_code: string;
  override_type: string;
  rule_reference: string | null;
  impact_summary: string;
  review_required: boolean;
  confidence_level: string;
  source_id: string | null;
  effective_from: string | null;
  effective_to: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface BuildingCommuneContext {
  building_id: string;
  city: string;
  canton: string;
  adapter: CommunalAdapter | null;
  overrides: CommunalOverride[];
}

export const swissRulesWatchApi = {
  async listSources(tier?: string): Promise<RuleSource[]> {
    const params = tier ? { tier } : {};
    const res = await apiClient.get<RuleSource[]>('/swiss-rules/sources', { params });
    return res.data;
  },

  async getUnreviewedChanges(): Promise<RuleChangeEvent[]> {
    const res = await apiClient.get<RuleChangeEvent[]>('/swiss-rules/changes/unreviewed');
    return res.data;
  },

  async getBuildingCommuneContext(buildingId: string): Promise<BuildingCommuneContext> {
    const res = await apiClient.get<BuildingCommuneContext>(`/buildings/${buildingId}/commune-context`);
    return res.data;
  },
};

File: backend/app/schemas/swiss_rules_watch.py
"""SwissRules Watch + Commune Adapter schemas."""

from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

# ---- RuleSource ----


class RuleSourceCreate(BaseModel):
    source_code: str
    source_name: str
    source_url: str | None = None
    watch_tier: str  # daily|weekly|monthly|quarterly
    is_active: bool = True
    notes: str | None = None


class RuleSourceRead(BaseModel):
    id: UUID
    source_code: str
    source_name: str
    source_url: str | None
    watch_tier: str
    last_checked_at: datetime | None
    last_changed_at: datetime | None
    freshness_state: str
    change_types_detected: list[str] | None
    is_active: bool
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---- RuleChangeEvent ----


class RuleChangeEventCreate(BaseModel):
    event_type: str  # new_rule|amended_rule|repealed_rule|portal_change|form_change|procedure_change
    title: str
    description: str | None = None
    impact_summary: str | None = None
    affects_buildings: bool = False


class RuleChangeEventRead(BaseModel):
    id: UUID
    source_id: UUID
    event_type: str
    title: str
    description: str | None
    impact_summary: str | None
    detected_at: datetime
    reviewed: bool
    reviewed_by_user_id: UUID | None
    reviewed_at: datetime | None
    review_notes: str | None
    affects_buildings: bool
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ReviewPayload(BaseModel):
    notes: str | None = None


# ---- CommunalAdapterProfile ----


class CommunalAdapterCreate(BaseModel):
    commune_code: str
    commune_name: str
    canton_code: str
    adapter_status: str = "draft"
    supports_procedure_projection: bool = False
    supports_rule_projection: bool = False
    fallback_mode: str = "canton_default"
    source_ids: list[str] | None = None
    notes: str | None = None


class CommunalAdapterRead(BaseModel):
    id: UUID
    commune_code: str
    commune_name: str
    canton_code: str
    adapter_status: str
    supports_procedure_projection: bool
    supports_rule_projection: bool
    fallback_mode: str
    source_ids: list[str] | None
    notes: str | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---- CommunalRuleOverride ----


class CommunalOverrideCreate(BaseModel):
    commune_code: str
    canton_code: str
    override_type: str
    rule_reference: str | None = None
    impact_summary: str
    review_required: bool = True
    confidence_level: str = "review_required"
    source_id: UUID | None = None
    effective_from: date | None = None
    effective_to: date | None = None
    is_active: bool = True


class CommunalOverrideRead(BaseModel):
    id: UUID
    commune_code: str
    canton_code: str
    override_type: str
    rule_reference: str | None
    impact_summary: str
    review_required: bool
    confidence_level: str
    source_id: UUID | None
    effective_from: date | None
    effective_to: date | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---- Building commune context (read-only projection) ----


class BuildingCommuneContext(BaseModel):
    building_id: UUID
    city: str
    canton: str
    adapter: CommunalAdapterRead | None
    overrides: list[CommunalOverrideRead]
